File: backend/app/models/leads.py
from __future__ import annotations

import uuid

from sqlalchemy import DateTime, ForeignKey, Index, Integer, String, Text
from sqlalchemy.dialects.postgresql import UUID as PGUUID
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base, TimestampMixin, UUIDPrimaryKeyMixin

LEAD_STATUSES = (
    "new",
    "qualifying",
    "qualified",
    "contacted",
    "appointment_booked",
    "application_started",
    "converted",
    "lost",
    "disqualified",
)
TERMINAL_LEAD_STATUSES = ("converted", "lost", "disqualified")
LEAD_TEMPERATURES = ("cold", "warm", "hot")
LEAD_INTENTS = ("informational", "exploring", "interested", "high_intent", "ready_to_apply")


class Lead(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    __tablename__ = "leads"
    __table_args__ = (
        # Serves analytics date-range queries (Task 013): every one
        # filters `college_id = ? AND created_at BETWEEN ? AND ?`.
        Index("ix_leads_college_created", "college_id", "created_at"),
    )

    college_id: Mapped[uuid.UUID] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("colleges.id"), nullable=False, index=True
    )
    student_id: Mapped[uuid.UUID] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("students.id"), nullable=False, index=True
    )
    course_id: Mapped[uuid.UUID | None] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("courses.id"), nullable=True
    )
    status: Mapped[str] = mapped_column(String(50), nullable=False, default="new", index=True)
    intent: Mapped[str | None] = mapped_column(String(50), nullable=True)
    lead_score: Mapped[int] = mapped_column(Integer, nullable=False, default=0, index=True)
    lead_temperature: Mapped[str | None] = mapped_column(String(20), nullable=True, index=True)
    source: Mapped[str] = mapped_column(String(50), nullable=False, default="voice_agent")
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    next_action: Mapped[str | None] = mapped_column(Text, nullable=True)
    hostel_interest: Mapped[bool | None] = mapped_column(nullable=True)
    scholarship_interest: Mapped[bool | None] = mapped_column(nullable=True)
    parent_involvement: Mapped[bool | None] = mapped_column(nullable=True)
    last_contacted_at: Mapped[DateTime | None] = mapped_column(DateTime(timezone=True), nullable=True)


class LeadScoreEvent(UUIDPrimaryKeyMixin, Base):
    __tablename__ = "lead_score_events"

    college_id: Mapped[uuid.UUID] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("colleges.id"), nullable=False, index=True
    )
    lead_id: Mapped[uuid.UUID] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("leads.id"), nullable=False, index=True
    )
    event_type: Mapped[str] = mapped_column(String(100), nullable=False)
    points: Mapped[int] = mapped_column(Integer, nullable=False)
    reason: Mapped[str | None] = mapped_column(Text, nullable=True)
    source: Mapped[str | None] = mapped_column(String(50), nullable=True)
    idempotency_key: Mapped[str | None] = mapped_column(String(255), nullable=True, index=True)
    created_at: Mapped[DateTime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default="now()"
    )

File: frontend/types/leads.ts
export type LeadTemperature = "hot" | "warm" | "cold";

export type LeadStatus =
  | "new"
  | "contacted"
  | "qualifying"
  | "qualified"
  | "appointment_booked"
  | "application_started"
  | "converted"
  | "lost"
  | "disqualified";

export type LeadStudent = {
  id: string;
  name: string | null;
  phone: string | null;
  email: string | null;
  qualification: string | null;
  qualification_score: number | null;
  entrance_exam: string | null;
  entrance_exam_score: number | null;
  budget: string | null;
  location: string | null;
  parent_name: string | null;
  parent_phone: string | null;
};

export type Lead = {
  id: string;
  college_id: string;
  student_id: string;
  course_id: string | null;
  course_name: string | null;
  status: LeadStatus | string;
  intent: string | null;
  lead_score: number;
  lead_temperature: LeadTemperature | string;
  source: string;
  notes: string | null;
  next_action: string | null;
  hostel_interest: boolean | null;
  scholarship_interest: boolean | null;
  parent_involvement: boolean | null;
  last_contacted_at: string | null;
  last_activity_at: string;
  created_at: string;
  updated_at: string;
  student: LeadStudent | null;
};

export type LeadListParams = {
  status?: string;
  temperature?: string;
  intent?: string;
  source?: string;
  sort?: "newest" | "recently_active" | "recently_updated" | "highest_score" | "lowest_score";
  page?: number;
  page_size?: number;
};

export type LeadScoreEvent = {
  id: string;
  lead_id: string;
  event_type: string;
  points: number;
  reason: string | null;
  source: string | null;
  created_at: string;
};

export type LeadScoreEvents = {
  lead_id: string;
  score: number;
  temperature: string;
  events: LeadScoreEvent[];
};
